### PR TITLE
feat(error): preserve error's source chain across gRPC boundary

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Recursively encode the error and its sources into into a _data_ struct, filling it into the `details` field of `tonic::Status`, then propagate it through the wire to the gRPC client and extract the source chain there with `TonicStatusWrapper`. This follows the practice of the [gRPC error handling guide](https://grpc.io/docs/guides/error/#richer-error-model).

Organize the utilities into a new crate `risingwave_error` to draw a clear distinction with the error handling pattern in `risingwave_common::error` that's going to be deprecated.

The idea goes back to https://github.com/risingwavelabs/risingwave/pull/4227#discussion_r931235861.

### Preview

https://github.com/risingwavelabs/risingwave/blob/c37d889fc91e16f4bc213c6ef771fd20dc7f4b0a/e2e_test/error_ui/main.slt#L13-L20

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/error.rs [hunk 1]
```diff
@@ -233,19 +233,19 @@ where
...
+            MetaErrorInner::PermissionDenied(_) => Code::PermissionDenied,
+            MetaErrorInner::CatalogIdNotFound(_, _) => Code::NotFound,
+            MetaErrorInner::Duplicated(_, _) => Code::AlreadyExists,
+            MetaErrorInner::Unavailable(_) => Code::Unavailable,
+            MetaErrorInner::Cancelled(_) => Code::Cancelled,
+            MetaErrorInner::InvalidParameter(_) => Code::InvalidArgument,
+            _ => Code::Internal,
+        };
+
+        err.to_status(code)
```
Comment (xxchan) [inline src/meta/src/error.rs:248]: cool 😮

File: src/error/src/tonic.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+pub struct TonicStatusWrapper(tonic::Status);
+
+impl TonicStatusWrapper {
+    /// Create a new [`TonicStatusWrapper`] from the given [`tonic::Status`] and extract
+    /// the source chain from its `details` field.
+    pub fn new(mut status: tonic::Status) -> Self {
+        if status.source().is_none() {
+            if let Ok(e) = bincode::deserialize::<serde_error::Error>(status.details()) {
+                status.set_source(Arc::new(e));
+            }
```
Comment (xxchan) [inline src/error/src/tonic.rs:55]: Is this failliabel? Maybe add a log line here.

File: src/common/src/error.rs [hunk 3]
```diff
@@ -195,12 +192,41 @@ pub struct RwError {
...
     fn from(err: RwError) -> Self {
-        match &*err.inner {
-            ErrorCode::ExprError(e) => tonic::Status::invalid_argument(e.to_string()),
-            ErrorCode::PermissionDenied(e) => tonic::Status::permission_denied(e),
-            ErrorCode::InternalError(e) => tonic::Status::internal(e),
-            _ => tonic::Status::internal(err.to_string()),
+        use tonic::Code;
+
+        let code = match &*err.inner {
+            ErrorCode::ExprError(_) => Code::InvalidArgument,
```
Comment (xxchan) [inline src/common/src/error.rs:198]: ExprError => InvalidArgument doesn't look very correct 🤣

BTW, I was always wondering whether we should map application error code to tonic status code.. This is like the debate whether to use HTTP status code or simply use it as a wrapper and do everything in the payload. I don't have an answer. Just want to mention it. Let's just keep it until somebody is unhappy.

File: src/error/src/tonic.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+impl From<tonic::Status> for TonicStatusWrapper {
+    fn from(status: tonic::Status) -> Self {
+        Self::new(status)
+    }
+}
+
+impl std::fmt::Display for TonicStatusWrapper {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "gRPC error ({}): {}", self.0.code(), self.0.message())
```
Comment (xxchan) [inline src/error/src/tonic.rs:79]: Why not being transparent here? It's bad? 🤣

File: src/error/src/tonic.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    /// The source chain is preserved by pairing with [`TonicStatusWrapper`].
+    // TODO(error-handling): disallow constructing `tonic::Status` directly with `new` by clippy.
+    pub fn to_status(&self, code: tonic::Code) -> tonic::Status {
+        // Embed the whole error (`self`) and its source chain into the details field.
+        // At the same time, set the message field to the error message of `self` (without source chain).
+        // The redundancy of the current error's message is intentional in case the client ignores the `details` field.
+        let source = serde_error::Error::new(self);
+        let details = bincode::serialize(&source).unwrap_or_default();
+
+        let mut status = tonic::Status::with_details(code, self.to_string(), details.into());
```
Comment (xxchan) [inline src/error/src/tonic.rs:34]: I'm not sure whether this is idiomatic. 

IIUC, `details` has a specific definition here:
- https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
- tonic's: https://docs.rs/tonic-types/latest/tonic_types/struct.ErrorDetails.html

We might need to put extra stuff in `metadata` 
- https://docs.rs/tonic-types/latest/tonic_types/trait.StatusExt.html

I'm not sure if the official `details` contains `source` and stuff.

**Review by xxchan — COMMENTED** [review]
Generally LGTM. Cool!

Could you please construct a real example in RisingWave where the err msg is improved? 🤪

File: src/error/Cargo.toml
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/error/Cargo.toml:1]: For me it's too similar to `common/src/error`. 😕

File: src/error/src/tonic.rs (hunk 4 above)
Comment (BugenZhao) [reply src/error/src/tonic.rs:79]: The `Display` implementation of `tonic::Status` is bad, and `#[error(transparent)]` does not allow us to customize the error message. 😕

File: src/common/src/error.rs (hunk 3 above)
Comment (BugenZhao) [reply src/common/src/error.rs:198]: I agree. I just leave the behavior unchanged in this PR. 😕

> whether we should map application error code to tonic status code.

+1. And the only place we refer to it is the error message of `Status`. 😄

File: src/error/src/tonic.rs (hunk 5 above)
Comment (BugenZhao) [reply src/error/src/tonic.rs:34]: Wow, didn't investigate this crate. Seems it's just the generated Rust code for the `ErrorDetails` message of gRPC standard.

> We might need to put extra stuff in `metadata`

Sure, I'll do that. Actually the `details` is also encoded in the metadata header with a reserved key.

**Comment by BugenZhao** [comment]
> Could you please construct a real example in RisingWave where the err msg is improved? 🤪

I'll show you after #13248 and https://github.com/risinglightdb/sqllogictest-rs/pull/200 🥰

File: e2e_test/error_ui/main.slt [hunk 6]
```diff
@@ -15,8 +15,9 @@ alter system set not_exist_key to value;
 ----
 db error: ERROR: QueryError
 
-Caused by this error:
-  1: internal error: SystemParams error: unrecognized system param "not_exist_key"
+Caused by these errors (recent errors listed first):
+  1: gRPC request failed: Internal error
+  2: SystemParams error: unrecognized system param "not_exist_key"
```
Comment (xxchan) [inline e2e_test/error_ui/main.slt:20]: Do you have any other more interesting example to demonstrate the change? 🤪 I feel this change isn't "aha" enough.

Comment (BugenZhao) [reply e2e_test/error_ui/main.slt:20]: > Do you have any other more interesting example to demonstrate the change? 🤪 I feel this change isn't "aha" enough.

I guess more examples will be in the developer logs (like actor bugs or network failure, not reviewed or refactored for now), as the frontend has cached the major stuff and done the check itself.

Comment (BugenZhao) [reply e2e_test/error_ui/main.slt:20]: > Besides, I'm wondering whether it's possible to have sth like `gRPC request to xx node failed` 🤪

Having the attempt. Stay tuned. 😄

**Review by xxchan — APPROVED** [review]

**Review by yezizp2012 — APPROVED** [review]

**Pull request merged**